src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java

```diff
@@ -0,0 +1,163 @@
+package sh.okx.railswitch;
+
+import java.util.Arrays;
+import java.util.Objects;
+import org.bukkit.Bukkit;
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.Sign;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.EntityType;
+import org.bukkit.entity.Minecart;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.BlockRedstoneEvent;
+import vg.civcraft.mc.citadel.Citadel;
+import vg.civcraft.mc.citadel.ReinforcementManager;
+import vg.civcraft.mc.citadel.reinforcement.PlayerReinforcement;
+import vg.civcraft.mc.citadel.reinforcement.Reinforcement;
+
+/**
+ * Event listener to enact the functionality of rail switches.
+ */
+public class DetectorRailActivateListener implements Listener {
+    
+    private static final String WILDCARD = "*";
+    
+    private final RailSwitchPlugin plugin;
+    
+    /**
+     * Creates a new event listener, passing through the current RailSwitch plugin instance.
+     *
+     * @param plugin The current RailSwitch plugin instance.
+     */
+    public DetectorRailActivateListener(RailSwitchPlugin plugin) {
+        this.plugin = plugin;
+    }
+    
+    /**
+     * Event handler for rail switches. Will determine if a switch exists at the target location, and if so will process
+     * it accordingly, allowing it to trigger or not trigger depending on the rider's set destination, the listed
+     * destinations on the switch, and the switch type.
+     *
+     * @param event The block redstone event to base the switch's existence on.
+     */
+    @EventHandler
+    public void onRailSwitch(BlockRedstoneEvent event) {
+        Block block = event.getBlock();
+        
+        // Block must be a detector rail being triggered
+        if (block == null
+                || block.getType() != Material.DETECTOR_RAIL
+                || event.getNewCurrent() != 15) {
+            return;
+        }
+        
+        // Check that the block above the rail is a sign
+        Block above = block.getRelative(BlockFace.UP);
+        if (above == null
+                // Do not check for Material.SIGN as that is an item material, not a block material
+                || (above.getType() != Material.SIGN_POST && above.getType() != Material.WALL_SIGN)
+                || !(above.getState() instanceof Sign)) {
+            return;
+        }
+        
+        // Check that the sign has a valid switch type
+        String[] lines = ((Sign) above.getState()).getLines();
+        SwitchType type = SwitchType.find(lines[0]);
+        if (type == null) {
+            return;
+        }
+        
+        // If Citadel is enabled, check that the sign and the rail are on the same group
```

Comment:
I'd do the player check before the reinforcement check to waste less time on empty carts (especially hopper carts are used a ton in some farms)

Reply:
Done